bftengine/src/bftengine/Bitmap.hpp

```diff
@@ -0,0 +1,146 @@
+      p_ = nullptr;
+    } else {
+      p_ = (unsigned char*)std::malloc(realSize());
+      std::memcpy(p_, other.p_, realSize());
+    }
+  }
+
+  ~Bitmap() {
+    if (numBits_ > 0) {
+      Assert(p_ != nullptr);
+      std::free(p_);
+    }
+  }
+
+  Bitmap& operator=(const Bitmap& other) {
+    if (numBits_ > 0) {
+      Assert(p_ != nullptr);
+      std::free(p_);
+    }
+
+    numBits_ = other.numBits_;
+    if (numBits_ == 0) {
+      p_ = nullptr;
+    } else {
+      p_ = (unsigned char*)std::malloc(realSize());
+      std::memcpy(p_, other.p_, realSize());
+    }
+
+    return *this;
+  }
+
+  void zeroAll() {
+    if (p_ == nullptr) return;
+    const uint32_t s = realSize();
+    Assert(s > 0);
+    memset((void*)p_, 0, s);
+  }
+
+  uint32_t numOfBits() const { return numBits_; }
+
+  bool get(uint32_t i) const {
+    Assert(i < numBits_);
+    const uint32_t byteIndex = i / 8;
+    const unsigned char byteMask = (1 << (i % 8));
+    return ((p_[byteIndex] & byteMask) != 0);
+  }
+
+  void set(uint32_t i) {
+    Assert(i < numBits_);
+    const uint32_t byteIndex = i / 8;
+    const unsigned char byteMask = (1 << (i % 8));
+    p_[byteIndex] = p_[byteIndex] | byteMask;
+  }
+
+  void reset(uint32_t i) {
+    Assert(i < numBits_);
+    const uint32_t byteIndex = i / 8;
+    const unsigned char byteMask = ((unsigned char)0xFF) & ~(1 << (i % 8));
+    p_[byteIndex] = p_[byteIndex] & byteMask;
+  }
+
+  void writeToBuffer(char* buffer,
+                     uint32_t bufferLength,
+                     uint32_t* actualSize) const {
+    const uint32_t sizeNeeded = sizeNeededInBuffer();
+    Assert(bufferLength >= sizeNeeded);
+    uint32_t* pNumOfBits = (uint32_t*)buffer;
+    char* pBitmap = buffer + sizeof(uint32_t);
+    *pNumOfBits = numBits_;
+    memcpy(pBitmap, p_, realSize());
+    if (actualSize) *actualSize = sizeNeeded;
+  }
+
+  uint32_t sizeNeededInBuffer() const {
+    return (sizeof(uint32_t) + realSize(numBits_));
+  }
+
+  static Bitmap* createBitmapFromBuffer(char* buffer,
+                                        uint32_t bufferLength,
+                                        uint32_t* actualSize) {
+    if (actualSize) *actualSize = 0;
+    if (bufferLength < sizeof(uint32_t)) return nullptr;
+    uint32_t* pNumOfBits = (uint32_t*)buffer;
+    const uint32_t numOfBitmapBytes = realSize(*pNumOfBits);
+    const uint32_t sizeNeeded = sizeof(uint32_t) + numOfBitmapBytes;
+    if (bufferLength < sizeNeeded) return nullptr;
+    Bitmap* b = new Bitmap();
+    if (*pNumOfBits > 0) {
+      char* pBitmap = buffer + sizeof(uint32_t);
+      b->numBits_ = *pNumOfBits;
+      b->p_ = (unsigned char*)std::malloc(numOfBitmapBytes);
+      std::memcpy(b->p_, pBitmap, numOfBitmapBytes);
+    }
+    if (actualSize) *actualSize = sizeNeeded;
+    return b;
+  }
+
+  static uint32_t maxSizeNeededToStoreInBuffer(uint32_t maxNumOfBits) {
+    return (sizeof(uint32_t) + realSize(maxNumOfBits));
+  }
+
+ protected:
```

Comment:
IMO these should default to private. Attempting to maintain invariants across inherited classes on protected fields when both the base and subclass can modify them is error prone in my experience.